mobile/src/main/java/org/openhab/habdroid/core/connection/ConnectionFactory.java

```diff
@@ -0,0 +1,282 @@
+        if (needInvalidateCachePreferenceKeys.contains(key)) {
+            cachedConnections.clear();
+        }
+    }
+
+    private void setSettings(SharedPreferences settings) {
+        if (this.settings != null) {
+            this.settings.unregisterOnSharedPreferenceChangeListener(this);
+        }
+        this.settings = settings;
+
+        this.settings.registerOnSharedPreferenceChangeListener(this);
+    }
+
+    public static Connection getConnection(int connectionType, Context ctx) {
+        return getConnection(connectionType, ctx,
+                PreferenceManager.getDefaultSharedPreferences(ctx));
+    }
+
+    public static Connection getConnection(int connectionType, Context ctx,
+                                           SharedPreferences settings) {
+        // FIXME: Once we're on API level 19, use Objects.requireNonNull();
+        if (ctx == null) {
+            throw new NullPointerException("Context ctx can not be null.");
+        }
+
+        ConnectionFactory factory = getInstance();
+        factory.setSettings(settings);
+        factory.setContext(ctx);
+
+        ctx.registerReceiver(factory, new IntentFilter(ConnectivityManager.CONNECTIVITY_ACTION));
+
+        if (factory.cachedConnections.containsKey(connectionType)) {
+            return factory.cachedConnections.get(connectionType);
+        }
+
+        Connection conn;
+        switch (connectionType) {
+            case Connection.TYPE_LOCAL:
+                conn = factory.getLocalConnection();
+                break;
+            case Connection.TYPE_REMOTE:
+                conn = factory.getRemoteConnection();
+                break;
+            case Connection.TYPE_CLOUD:
+                // TODO: Need a proper way of finding if the connection supports openHAB cloud
+                // things, e.g. by checking if the /api/v1/settings/notifications endpoint works,
+                // but currently does not work for myopenhab.org
+                conn = factory.getRemoteConnection();
+                break;
+            case Connection.TYPE_ANY:
+                conn = factory.getAvailableConnection();
+                break;
+            default:
+                throw new IllegalArgumentException("Invalid Connection type requested.");
+        }
+
+        if (conn != null)
+            factory.cachedConnections.put(connectionType, conn);
+
+        return conn;
+    }
+
+    private Connection getLocalConnection() {
+        String openHABUrl = Util.normalizeUrl(settings.getString(Constants
+                .PREFERENCE_LOCAL_URL, ""));
+        // If local URL is configured
+        if (openHABUrl.length() > 0) {
+            Log.d(TAG, "Connecting to local URL " + openHABUrl);
+            Connection conn = new DefaultConnection(ctx);
+            conn.setOpenHABUrl(openHABUrl);
+            conn.setConnectionType(Connection.TYPE_LOCAL);
+
+            if (hasUsername()) {
```

Comment:
Why are those checks needed here? If username is null, the setter should be a no-op, shouldn't it?

Also, can't all those setters be removed from the `Connection` interface and be inlined into the respective constructor as constructor parameters?

Reply:
Possible, I don't care that much -> changed :)